### PR TITLE
More NCBI changes for your consideration (Jan. 2022)

### PR DESCRIPTION
**Pull request by ucko**
Happy new year!  I have four more patches for your consideration; I've rolled them up into a single PR, but they're mutually independent, so you should be able to take or leave them individually.  (Likewise for 27c7a831's changes to each individual file.)

### CODE REVIEW

File: src/ctlib/unittests/rpc_ct_setparam.c [hunk 1]
```diff
@@ -414,14 +415,14 @@ int i, j;
 				coldata[i].value[0] = 0;
 				if (coldata[i].value == NULL) {
 					fprintf(stderr, "malloc coldata.value failed \n");
-					return 1;
+					break;
```
Comment (fziglio) [inline src/ctlib/unittests/rpc_ct_setparam.c:418]: I think you need also to set ret to `CS_FAIL`, otherwise you won't report an error

File: src/tds/convert.c [hunk 2]
```diff
@@ -1530,13 +1530,13 @@ tds_convert_real(const TDS_REAL* src, int desttype, CONV_RESULT * cr)
 		return sizeof(TDS_USMALLINT);
 		break;
 	case SYBINT4:
-		if (!IS_INT(the_value))
+		if (!IS_INT((double)the_value))
```
Comment (fziglio) [inline src/tds/convert.c:1533]: nice catch, not sure however if this is really fixing an issue or just hiding a warning. It fixes loosing precision in the conversion but is it possible that it does not catch a number which could not be converted to int, or, otherwise, that it returns overflow for a number that can be converted to int?

File: src/ctlib/unittests/rpc_ct_setparam.c (hunk 1 above)
Comment (ucko) [reply src/ctlib/unittests/rpc_ct_setparam.c:418]: I'll take another look, thanks.  Same goes for `lang_ct_param.c`, I suppose?

File: src/tds/convert.c (hunk 2 above)
Comment (ucko) [reply src/tds/convert.c:1533]: As https://godbolt.org/z/Ea18E5zTM illustrates, the issue is that due to limited `float` precision, you can't get too close to the limit without going slightly *over* it, at which point conversion will wrap around rather than saturating.

Comment (fziglio) [reply src/tds/convert.c:1533]: It's more about rounding and precision. I was just wondering what's the best way without having to do conversions. Also I was looking at same function trying to understand if the same issue could affect other conversions.

**Comment by fziglio** [comment]
Merged
- ` _ct_handle_client_message: Report appropriate severity values` (minor space changes)
- `ct_describe: For CS_BIGDATETIME_TYPE, supply the server type in usertype`

File: src/ctlib/unittests/rpc_ct_setparam.c (hunk 1 above)
Comment (fziglio) [reply src/ctlib/unittests/rpc_ct_setparam.c:418]: > I'll take another look, thanks. Same goes for `lang_ct_param.c`, I suppose?

where exactly?

File: src/tds/convert.c (hunk 2 above)
Comment (ucko) [reply src/tds/convert.c:1533]: Good question.  Upon further consideration, how about leaving these call sites alone and instead tweaking the macros?  AFAICT,
```
#define IS_INT(x) (TDS_INT_MIN <= (x) && (x) < (int64_t) TDS_INT_MAX + 1)
/* ... */
#define IS_UINT(x) (0 <= (x) && (x) < (uint64_t) TDS_UINT_MAX + 1)
```
should impose precisely correct limits.  They will let through `double` values exceeding `TDS_(U)INT_MAX` by less than one, but I'd argue that that's actually appropriate given that conversion is defined to round down anyway.

Comment (fziglio) [reply src/tds/convert.c:1533]: Agreed. I would put all signed in the macros like
```c
#define IS_INT(x) (TDS_INT_MIN <= (x) && (x) < (int64_t) TDS_INT_MAX + 1)
/* ... */
#define IS_UINT(x) (0 <= (x) && (x) < (int64_t) TDS_UINT_MAX + 1)
```
not sure about `int64_t`/`TDS_INT8`, maybe a future style update for the other parts

Comment (ucko) [reply src/tds/convert.c:1533]: I'm on board with the all-signed version.  64-bit types are trickier, particularly on the unsigned front, but AFAICT the only use of `IS_(U)INT8` is on a `long` (line 1116), where it amounts to a simple non-negativity check -- `tds_convert_real` and `tds_convert_flt8` use hand-coded boundary checks that attempt to test directly against `_MAX` values.  I'll prepare a revised patch.

File: src/ctlib/unittests/rpc_ct_setparam.c (hunk 1 above)
Comment (ucko) [reply src/ctlib/unittests/rpc_ct_setparam.c:418]: Never mind, I misremembered which pair of tests wound up having parallel fixes.

File: src/tds/convert.c [hunk 3]
```diff
@@ -1538,13 +1538,13 @@ tds_convert_real(const TDS_REAL* src, int desttype, CONV_RESULT * cr)
...
 	case SYBINT8:
-		if (the_value > (TDS_REAL) TDS_INT8_MAX || the_value < (TDS_REAL) TDS_INT8_MIN)
+		if (!IS_INT8(the_value))
 			return TDS_CONVERT_OVERFLOW;
 		cr->bi = (TDS_INT8) the_value;
 		return sizeof(TDS_INT8);
 		break;
 	case SYBUINT8:
-		if (the_value > (TDS_REAL) TDS_UINT8_MAX || the_value < 0)
+		if (the_value >= TDS_UINT8_MAX + 1.0f || the_value < 0)
```
Comment (fziglio) [inline src/tds/convert.c:1547]: Wondering if this `TDS_UINT8_MAX + 1.0f` is portable or not. Technically the compiler will convert the first number to float then do the addition. However nor float nor double have the precision to hold all the number precisely (double has 54 bit precision while he number is 63 bit precise). It will probably round directly the number, then add 1 (which won't change). Potentially the result is what we are expecting, but hard to say. If the compile would do truncation on the first conversion the result is not the expected number.

Comment (fziglio) [reply src/tds/convert.c:1547]: Why not having something like
```c
#define TDS_UINT8_UPPER 18446744073709551616.0
```

File: src/tds/convert.c [hunk 4]
```diff
@@ -78,23 +78,23 @@ static const char *parse_numeric(const char *buf, const char *pend,
...
+#define IS_INT(x) (TDS_INT_MIN <= (x) && (x) < (int64_t) TDS_INT_MAX + 1)
 
 #define TDS_UINT_MAX 4294967295lu
-#define IS_UINT(x) (0 <= (x) && (x) <= TDS_UINT_MAX)
+#define IS_UINT(x) (0 <= (x) && (x) < (int64_t) TDS_UINT_MAX + 1)
 
 #define TDS_INT8_MAX ((((TDS_INT8) 0x7fffffffl) << 32) + (TDS_INT8) 0xfffffffflu)
 #define TDS_INT8_MIN (-(((TDS_INT8)1)<<62) -(((TDS_INT8)1)<<62))
-#define IS_INT8(x) (TDS_INT8_MIN <= (x) && (x) <= TDS_INT8_MAX)
+#define IS_INT8(x) (TDS_INT8_MIN <= (x) && (x) < (uint64_t) TDS_INT8_MAX + 1)
```
Comment (fziglio) [inline src/tds/convert.c:97]: It should be fine... but `IS_INT8(-1)` is false. It's fine in the sense that we should use this to check unsigned numbers and floating point where the check is correct. But if this is used for signed numbers it will fail.

**Comment by fziglio** [comment]
Merged static analyzer commit. I tried to fix the possible truncating issue, see https://github.com/freddy77/freetds/commit/b0b8887ce6d254a6e04773205fa6a71cc13a87b7 (on top of your, potentially a fixup commit).

File: src/tds/convert.c (hunk 3 above)
Comment (ucko) [reply src/tds/convert.c:1547]: From what I gather (details below), the `TDS_UINT8_MAX` approach should be pretty portable.  Upon consideration, though, a third possibility could be to take advantage of the fact that `TDS_UINT8_MIN` should have an exact representation by multiplying it by `-2.0` -- per https://stackoverflow.com/questions/29174165/is-flt-radix-ever-not-2-in-c11-for-modern-general-purpose-computers, in the rare event that `FLT_RADIX` isn't two, it should still be a power of two, which suffices (with decimal floating-point, if supported, having separate `_DecimalN` types).

As for rounding, per https://en.cppreference.com/w/c/language/conversion#Real_floating-integer_conversions, on implementations supporting IEEE floating point (which should be more or less all of them), rounding is to nearest for anything evaluable at compile time unless `FENV_ACCESS` (https://en.cppreference.com/w/c/preprocessor/impl) is `ON` (unusual) *and* the application has used `fesetround` to specify some other rounding mode.

Meanwhile, I'd forgotten that the first step in conversion from floating point was truncation rather than taking a floor, so I'll want to allow more margin on signed lower bounds accordingly.

File: src/tds/convert.c (hunk 4 above)
Comment (ucko) [reply src/tds/convert.c:97]: Good point.  I suppose an explicit comparison with zero is in order, though doing it up front should at least keep the total number of comparisons at two:
`#define IS_INT8(x) (((x) < 0) ? TDS_INT8_MIN <= (x) : (x) < (uint64_t) TDS_INT8_MAX + 1)`

File: src/tds/convert.c [hunk 5]
```diff
@@ -78,26 +78,37 @@ static const char *parse_numeric(const char *buf, const char *pend,
...
-#define IS_USMALLINT(x) ( 0 <= (x) && (x) <= 65535 )
+/*
+ * Confirm convertability to a signed type with the given bounds,
+ * bearing in mind that promotion rules can complicate comparison
+ * between negative and unsigned integers, that conversion from
+ * floating-point values to integers starts by truncating fractional
+ * portions, and that lb - 1 might have the same FP representation as
+ * lb in x's type (as with float -> int and anything -> TDS_INT8).
+ */
+#define IS_INTn(x, lb, ub) ((x) < 0 ? lb - (x) < 1 : (x) < ub + 1)
```
Comment (fziglio) [inline src/tds/convert.c:89]: To be honest this looks complicated... and also I don't think compilers can reach the level of the 2 initial simple comparisons. And the `lb - (x)` part could be subject to loose of precision.

File: src/tds/convert.c [hunk 6]
```diff
@@ -78,26 +78,37 @@ static const char *parse_numeric(const char *buf, const char *pend,
...
-#define TDS_INT8_MAX ((((TDS_INT8) 0x7fffffffl) << 32) + (TDS_INT8) 0xfffffffflu)
+#define TDS_INT8_MAX ((((TDS_INT8) 0x7fffffff) << 32) + (TDS_INT8) 0xffffffffu)
 #define TDS_INT8_MIN (-(((TDS_INT8)1)<<62) -(((TDS_INT8)1)<<62))
-#define IS_INT8(x) (TDS_INT8_MIN <= (x) && (x) <= TDS_INT8_MAX)
+#define IS_INT8(x) IS_INTn(x, TDS_INT8_MIN, (TDS_UINT8) TDS_INT8_MAX)
 
-#define TDS_UINT8_MAX ((((TDS_UINT8) 0xfffffffflu) << 32) + 0xfffffffflu)
+#define TDS_UINT8_MAX ((((TDS_UINT8) 0xffffffffu) << 32) + 0xffffffffu)
 #define IS_UINT8(x) (0 <= (x) && (x) <= TDS_UINT8_MAX)
+#define FLOAT_IS_UINT8(x) ((x) < -2.0f * TDS_INT8_MIN)
```
Comment (fziglio) [inline src/tds/convert.c:111]: I think you forgot to check for negativity.

**Comment by fziglio** [comment]
Overall I don't know what to suggest. Macros are useful if easy and reused, but if they have so much side effects expanded in a context or in another I don't know if they make much sense at this point. Another thing I would consider is a test to check the various boundaries.

File: src/tds/convert.c (hunk 6 above)
Comment (ucko) [reply src/tds/convert.c:111]: Oops, so I did.

**Comment by ucko** [comment]
Yeah, a test is probably a good idea; should it go to the existing `src/tds/unittests/convert.c`, or should I introduce a `convert2.c` or the like?

File: src/tds/convert.c (hunk 5 above)
Comment (ucko) [reply src/tds/convert.c:89]: The general idea with this macro is to unify cases that would otherwise require slightly different pairs of comparisons.  In particular:
- For `int`, `(x) >= TDS_INT_MIN` is correct for every source type but `double`, where it misses the open interval `(TDS_INT_MIN - 1, TDS_INT_MIN)`, and `(x) > (TDS_INT8) TDS_INT_MIN - 1` is correct for every source type but `float`, where `TDS_INT_MIN - 1` is unrepresentable and typically rounds to `TDS_INT_MIN`.  Looking at `lb - (x)` should yield correct results in either case, but will misfire if `x` is `unsigned int` or `TDS_UINT8`, necessitating upfront comparison with zero to be fully general.
- For `TDS_INT8`, `(x) <= TDS_INT8_MAX` picks up floating-point values that are slightly out of range and will wrap around upon conversion; `(x) < (TDS_UINT8) TDS_INT8_MAX + 1` fixes that but proceeds to reject negative integer values, likewise necessitating upfront comparison with zero to be fully general.

At any rate, I'm comparing `lb - (x) < 1` rather than `lb - 1 < (x)` due to precision concerns -- if `lb - 1` loses precision, the comparison could (and typically would) wrongly exclude `lb`, whereas `lb - (x)` will lose precision only if the difference is very positive (and hence clearly greater than one) or very negative (and hence clearly less than one).

**Comment by fziglio** [comment]
I finally manage to write the test. See https://github.com/FreeTDS/freetds/commit/ba2c5e2a152c6b980cd1beebba6f5947acc89510 and https://github.com/FreeTDS/freetds/commit/1ad2a894d4b0c1e8d1a6e3b7a282e73005101d9c.

**Comment by ucko** [comment]
Nice, thanks!  This test passes with my previously submitted changes plus that missing comparison with zero; I've rebased my branch accordingly and pushed it.

File: src/tds/convert.c (hunk 5 above)
Comment (fziglio) [reply src/tds/convert.c:89]: I tried something similar under compiler explorer:
```c++
#define INSIDE(x) ((x) > -4 && (x) < 10)

bool limit1(int a) {
    return INSIDE(a);
}

bool limit2(double a) {
    return INSIDE(a);
}

#define IS_INTn(x, lb, ub) ((x) < 0 ? lb - (x) < 1 : (x) < ub + 1)
#define INSIDE2(a) IS_INTn(a, -4, 10)

bool limit3(int a) {
    return INSIDE2(a);
}

bool limit4(double a) {
    return INSIDE2(a);
}
```
and in all compilers I tried the code is less efficient than normal bound checks (as I expected).
I wrote this https://github.com/freddy77/freetds/commit/7ebc5d85e6c001838bd2f6cba4bd623f5e5eea48 to simplify and use simpler bound checks. Please consider to squash it in.

Comment (ucko) [reply src/tds/convert.c]: Squashed, thanks, with some fine-tuning to account for pre-conversion truncation of negative FP values (which I'd previously missed myself for unsigned destination types).

**Comment by fziglio** [comment]
Merged. Wondering however if the test is completely testing correctly bounds for floating points

**Comment by ucko** [comment]
Molto grazie!  On the test front, perhaps you could make judicious use of `nextafter` and `nextafterf`.  On the upper end, the first out-of-range value will be some (exactly representable) power of two `x`, with `nextafter[f](x, 0)` barely making the cut.  The lower end for unsigned types is likewise straightforward -- `-1` is out, but `nextafter[f](-1, 0)` should be legitimate.  For signed types, it's alas not quite so simple -- if `x` is a floating point value corresponding to the lowest in-range integer (the negative of a power of two), then you can check whether (e.g.) `x-1 == x` to determine whether the cutoff is between `nextafter[f](x, 2*x)` and `x` or between `x-1` and `nextafter[f](x-1, 0)`.
